### DISCUSSION TITLE
Selector in CLI is ignored

### DISCUSSION THREAD

**Discussion by PHPirat** [deployphp/deployer#3950] [Help needed] [answered]
Hello,

I'm switching from Deployer v6 to v7.4.1.  I did follow the [official migration guide](https://deployer.org/docs/7.x/UPGRADE#upgrade-from-6x-to-7x) and ended up with this `deploy.php`:

```
<?php

set( 'default_selector', 'stage=staging' );
set( 'application', 'test' );
set( 'user', 'example' );

$stages = [
	'production' => 'master',
	'staging' => 'development'
];

array_walk( $stages, function( $branch, $stage ) {
	$serverCount = "[01:05]";
	host( "node-{$serverCount}.example.com/{$stage}" )
		->setLabels( [ 'stage' => $str_stage ] )
		->setRemoteUser( get( 'user' ) )
		->setForwardAgent( true )
		->setSshMultiplexing( true )
		->setSshArguments( [
			'-o UserKnownHostsFile=/dev/null',
			'-o StrictHostKeyChecking=no'
		] )
		->set( 'branch', function() use ( $branch ) {
			return input()->getOption( 'branch' ) ?: $branch;
		} )
		->set( 'deploy_path', "/var/www/{{application}}/{$stage}" );
} );
```
When running the command `php deployer.phar deploy stage=production` it deploys to the staging host.
If I set the `default_selector` to a non-existing value like `abcd` and run the same command again, i do get the error `No host selected. Please, check your selector: stage=abcd` (as expected). So I do assume that my host definitions are correct (deploying to staging is working fine).

Can somebody point me towards my mistake? 

On side note, the migration guide does not mention that the `RuntimeException` was renamed to `RunException`. I do think that this should be included in the guide, at least it took me some time to figure it out on my own.

**Comment by antonmedv** [comment]
> On side note, the migration guide does not mention that the RuntimeException was renamed to RunException. I do think that this should be included in the guide, at least it took me some time to figure it out on my own.

Good point.

> Can somebody point me towards my mistake?

I'm not completely getting but your mistake is. Could you please explain and show some examples?

**Reply by PHPirat** [reply]
I'm sorry for being unclear.

Previously, in v6, we used `host(...)->stage( $stage )` to set the stage and then run `php deployer.phar deploy staging` to deploy to our staging host or `php deployer.phar deploy production` to deploy to production.

As instructed in the migration guide, I did switch out `->stage( $str_stage )` for `->setLabels( [ 'stage' => $stage ] )` (4.iii.) and used `set( 'default_selector', 'stage=staging' );` instead of `set( 'default_stage', 'staging' );` (13.)

With those changes, I would've expected it to behave in the same way. But no matter what parameter I add to the `deploy` command, it won't get recognized as a selector and thus the `default_selector` value is used.
- `php deployer.phar deploy stage=staging` will deploy to staging correctly (but not if I change the `default_selector` to something non-existing)
- `php deployer.phar deploy stage=production` will also deploy to the staging instead of the production host

Passing other arguments like `--branch=htx/fix-error` or `-vvv` do work.

I only recently got into this topic so I'm still trying to wrap my head around it and I hope I made my problem clear.

PS
I just noticed that there is a small error in the `deploy.php` above.
`->setLabels( [ 'stage' => $str_stage ] )` should be `->setLabels( [ 'stage' => $stage ] )`, but this mistake is not in the actual file.

**Reply by antonmedv** [reply]
What is $str_stage?

**Reply by PHPirat** [reply]
Ok, so first of all, I do think I found the problem:
It seems that setting the `default_selector` overrides the selector set in the command. Removing it does solve the problem and the correct host it selected. If I don't supply any selector to the command, I'm asked to chose a host as expected.
What I still don't get is why it overrides the command input instead of being a fallback?

Regarding your question:
`$str_stage` or `$stage` as it should be called, is the value passed by `array_walk` function. The different stages are defined inside `$stages` where the key is equal to the stage and the value is the default branch to use. `$stage` is therefore equal to either "staging" or "production".
Our stages mainly differ by the deploy-path, that's why we used `array_walk` to define both stages inside the loop.

**Reply by antonmedv** [reply]
> What I still don't get is why it overrides the command input instead of being a fallback?

A think it is a bug.

**Reply by PHPirat** [reply]
Do you think the problem is on my side or is it more likely a general bug?

**Reply by antonmedv** [reply]
I need to check it. You can also try to reproduce the error with simple deploy.php.

**Reply by uacode** [reply]
@antonmedv Hello, seems to me, the dynamic Labels does not work
```php
->setLabels([
//        'stage' => 'prod',
        'stage' => function () {
            return getHost() === 'vps-2.domain.net' ? 'prod' : 'dev';
        },
        'role' => 'web',
    ]);
```

**Comment by florinr08** [comment]
What's taking so long to find and fix the issue here @antonmedv? It's been more than 2 months since this issue was raised and there's still no fix! The specified selector and `default_selector` act like default and override respectively.

Using a simple deploy.php file:
```php
set('default_selector', 'type=web');

host('web.example.com')
    ->setLabels([
        'type' => 'web',
        'env' => 'prod',
    ]);

host('db.example.com')
    ->setLabels([
        'type' => 'db',
        'env' => 'prod',
    ]);

task('info', function () {
    writeln('type:' . get('labels')['type'] . ' env:' . get('labels')['env']);
});
```

If you run this command:
```bash
$ dep info all
task info
[web.example.com] type:web env:prod

$ dep info anythingyoulikehere
task info
[web.example.com] type:web env:prod

$ dep info this_gets_completely_ignored_when_default_selector_is_set
task info
[web.example.com] type:web env:prod
```

The `default_selector` config is acting like the supplied input parameter that overrides any defaults set. Any selector I provide in my command will be ignored because there's a `default_selector` config set that will always override.

-------

The issue is here: https://github.com/deployphp/deployer/blob/v7.5.8/src/Command/SelectCommand.php#L54

```php
$selector = Deployer::get()->config->get('default_selector', $input->getArgument('selector'));
```

Yes, you're wanting to set a default for _something_ but you're setting the wrong default...

This should instead read something like this:

```php
$selector = $input->getArgument('selector');
$selector = empty($selector) ? Deployer::get()->config->get('default_selector') : $selector;
```

**Reply by antonmedv** [reply]
Make sense. I’m planning to fix it next September.  =) But you can always motivate me by supporting via GitHub Sponsors. Let say for one time donation of $500 I will fix it today!

**Comment by antonmedv** [comment, marked as answer]
Fixed in v7.5.10 and v8-alpha.1.
